plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ModuleUtils.java

```diff
@@ -56,4 +60,27 @@ public static String getServiceId(IModule module) {
     
     return "default";
   }
+
+  /**
+   * Returns the set of all referenced modules, including child modules. This returns the unique
+   * modules, and doesn't return the module paths.
+   */
+  public static IModule[] getAllModules(IServer server) {
+    Set<IModule> modules = new LinkedHashSet<>();
+    for (IModule module : server.getModules()) {
+      modules.add(module);
+      addChildModules(server, new IModule[] {module}, modules);
+    }
+    return modules.toArray(new IModule[modules.size()]);
+  }
+
+  /** Recursively walk the children from {@code modulePath}. */
+  private static void addChildModules(IServer server, IModule[] modulePath, Set<IModule> modules) {
+    IModule[] newModulePath = Arrays.copyOf(modulePath, modulePath.length + 1);
+    for (IModule child : server.getChildModules(modulePath, null)) {
```

Comment:
This looks like it may have combinatorial explosion. I.e. every child module is going to check all the modules in modulePath again rather than just the newly added module. Or is there some way this doesn't happen?